### ISSUE TITLE
Issues with inline array types in method bodies

### ISSUE THREAD

**Issue by DavideRossi** [mulesoft-labs/raml-for-jax-rs#337]
To my understanding this is a valid RAML method declaration:
```
    get:
      displayName: find
      responses:
        200:
          body:
            type: array
            items: RecipeReference
```
but this seems to trigger a bug in raml-to-jax-rs since a weird `List<org.foo.bar.generated.model.Object> entity` parameter is produced in the corresponding Java method.

If I declare the array as a type:
```
RecipeReferenceArray:
    type: array
    items: RecipeReference
```
and convert the method to:
```
    get:
      displayName: find
      responses:
        200:
          body:
            type: RecipeReferenceArray
```

the java method correctly returns `List<RecipeReference> entity`.

This is a full version of the RAML spec:

```
#%RAML 1.0
---
title: RecipesApi
baseUri: http://myrecipes.org/api
version: 1
mediaType: application/json

types:
  RecipeReference:
    type: object
    properties:
      name:
        type: string
      uri:
        type: string
  RecipeReferenceArray:
    type: array
    items: RecipeReference

/recipes:
    displayName: RecipeCollectionResource
    get:
      displayName: find
      responses:
        200:
          body:
            type: RecipeReferenceArray

```

**Comment by jpbelang** [comment]
Are you on the 3.0 branch ?

**Comment by DavideRossi** [comment]
No, I'm on 2.1.

**Comment by jpbelang** [comment]
Guess what my next question is :-) ?

Can you check 3.0.0-SNAPSHOT ?

**Comment by DavideRossi** [comment]
OK, I think I have a guess.
But I have a problem too: I was ready to distribute my uml2raml generator and I was writing a tutorial on how to use it with raml-to-jax-rs and I want to point to software that is available on maven central. Any hope to see a fix backported to the v2 branch, or any relase date for 3.0.0?
BTW: are you asking me to test that with v3 or do you know already it is going to work?

**Comment by jpbelang** [comment]
Release date for 3.0 is this weekend, pending if it solves your problem.

As to the BTW:  the inline types were made better in 3.0.  I can check your RAML this PM.  For the record, inline types are difficult to figure out, as the parser gives us no help on the subject, so we are left at guessing whether this is a new type.

**Comment by DavideRossi** [comment]
To me inline types should be ditched, they are hard to parse and are hard to read. But in my case I would have to create new synthetic type each time a method parameter is an array in the UML model. Testing it with 3.0.0 now...

**Comment by DavideRossi** [comment]
OK, same problem with 3.0.0-SNAPSHOT, the method is a little different but the parameter is still List<org.foo.bar.generated.model.Object> entity

**Comment by jpbelang** [comment]
I'll look.  I've seen this.  It might have to do with the raml.  This PM, for sure.

**Comment by DavideRossi** [comment]
And, BTW, in the case the inline declaration is in the request body the problem in a spurious `import org.foo.bar.generated.model.Object;`

**Comment by jpbelang** [comment]
Ok, here is the quick analysis:
Fails (array.items().name() == object, array.items().type() == RecipeReference):
```yaml
        200:
          body:
            type: array
            items: RecipeReference
```
SucceedsFails (array.items().name() == RecipeReference, array.items().type() == object):
```yaml
        200:
          body:
            type: RecipeReference[]
```
Which really surprises me, as it's semantically identical.  IMO.

But the fact that this succeeds:
```yaml
types:
  RecipeReferenceArray:
    type: array
    items: RecipeReference
```
Is the surprising part.  It should return the same as the first one.
So i fixed this like so (in the ramltopojo project). This should work for me as I don't create inline data types:
```java
        if ( "object".equals(itemTypeName)) {
            itemTypeName = typeDeclaration.items().type();
        }

        if ( "object".equals(itemTypeName)) {

            throw new GenerationException("unable to create type array item of type object (or maybe an inline array type ?)");
        }
```

Can you make sure it works for you ?

**Comment by jpbelang** [comment]
For the record:
https://github.com/raml-org/raml-java-parser/issues/250

**Comment by DavideRossi** [comment]
Yup, it works now (I had to delete ~/.m2/repository/org/raml, I do really really love Maven, right?).
So the question is: is this going to be backported to 2.1? I already wrote a workaround (producing a new <type>Array type for all arrays in method parameters and returns) and I want to know if I have to keep it for those working with 2.1 or I can safely ditch it (it's ugly!).

**Comment by jstoiko** [comment]
This won't be back-ported. We are about to release 3.0.0 though.
